diff -ruN parsec-2.0-orig/pkgs/apps/x264/src/encoder/encoder.c parsec-2.0/pkgs/apps/x264/src/encoder/encoder.c
--- parsec-2.0-orig/pkgs/apps/x264/src/encoder/encoder.c	2013-02-07 14:42:54.880848799 -0500
+++ parsec-2.0/pkgs/apps/x264/src/encoder/encoder.c	2013-02-09 01:16:08.535421000 -0500
@@ -31,6 +31,7 @@
 #include "analyse.h"
 #include "ratecontrol.h"
 #include "macroblock.h"
+#include "tern/user.h"
 
 #if VISUALIZE
 #include "common/visualize.h"
@@ -1176,6 +1177,7 @@
          * Slice P: choose between using P mode or intra (4x4 or 16x16)
          * */
         x264_macroblock_analyse( h );
+	tern_lineup(0);// The x264_macroblock_analyse and x264_fdec_filter_row contain sync op, so we have to put it here.
 
         /* encode this macroblock -> be careful it can change the mb type to P_SKIP if needed */
         x264_macroblock_encode( h );
diff -ruN parsec-2.0-orig/pkgs/apps/x264/src/x264.c parsec-2.0/pkgs/apps/x264/src/x264.c
--- parsec-2.0-orig/pkgs/apps/x264/src/x264.c	2013-02-07 14:42:54.880848799 -0500
+++ parsec-2.0/pkgs/apps/x264/src/x264.c	2013-02-09 01:17:03.703562000 -0500
@@ -32,6 +32,7 @@
 #include "common/cpu.h"
 #include "x264.h"
 #include "muxers.h"
+#include "tern/user.h"
 
 #ifndef _MSC_VER
 #include "config.h"
@@ -126,7 +127,7 @@
 
     /* Control-C handler */
     signal( SIGINT, SigIntHandler );
-
+	tern_lineup_init(0, param.i_threads, param.i_threads*50);
     ret = Encode( &param, &opt );
 
 #ifdef PTW32_STATIC_LIB
