diff -Naur gcc-4.2-src/libgomp/config/posix/bar.c gcc-4.2-src_modified/libgomp/config/posix/bar.c
--- gcc-4.2-src/libgomp/config/posix/bar.c	2013-01-17 17:45:24.804248297 -0500
+++ gcc-4.2-src_modified/libgomp/config/posix/bar.c	2013-01-21 10:50:01.856359610 -0500
@@ -31,6 +31,7 @@
    POSIX pthread_barrier_t won't work.  */
 
 #include "libgomp.h"
+#include "tern/user.h"
 
 
 void
@@ -44,6 +45,7 @@
   gomp_sem_init (&bar->sem2, 0);
   bar->total = count;
   bar->arrived = 0;
+  tern_lineup_init((long)bar, count, 100);
 }
 
 void
@@ -108,4 +110,6 @@
 gomp_barrier_wait (gomp_barrier_t *barrier)
 {
   gomp_barrier_wait_end (barrier, gomp_barrier_wait_start (barrier));
+  tern_lineup_start((long)barrier);
+  tern_lineup_end((long)barrier);
 }
