diff -ruN ImageMagick-6.8.0-4/magick/magick.c ImageMagick-6.8.0-4-modified/magick/magick.c
--- ImageMagick-6.8.0-4/magick/magick.c	2012-01-17 22:01:56.000000000 -0500
+++ ImageMagick-6.8.0-4-modified/magick/magick.c	2012-12-30 00:34:45.271586227 -0500
@@ -79,6 +79,7 @@
 #include "magick/token.h"
 #include "magick/utility.h"
 #include "magick/xwindow-private.h"
+#include <tern/user.h>
 
 /*
   Define declarations.
@@ -1216,6 +1217,7 @@
     *events,
     execution_path[MaxTextExtent],
     filename[MaxTextExtent];
+  soba_init(0, omp_get_max_threads(), omp_get_max_threads()*20);
 
   /*
     Initialize the Magick environment.

