--- pbzip2-0.9.5/pbzip2.cpp	2005-12-31 13:41:42.000000000 -0500
+++ pbzip2-0.9.5-modified/pbzip2.cpp	2013-03-26 01:45:35.756194715 -0400
@@ -87,6 +87,7 @@
 #include <time.h>
 #include <utime.h>
 #include <bzlib.h>
+#include <tern/user.h>
 #ifndef WIN32
 #include <sys/time.h>
 #include <unistd.h>
@@ -649,6 +650,7 @@
 			fprintf(stderr, " *ERROR: Could not allocate memory (DecompressedData)!  Skipping...\n");
 			return (NULL);
 		}
+		tern_lineup(0);
 
 		// decompress the memory buffer (verbose=0)
 		ret = BZ2_bzBuffToBuffDecompress(DecompressedData, &outSize, FileData, inSize, 0, Verbosity);
@@ -674,6 +676,7 @@
 				return (NULL);
 			}
 
+			tern_lineup(0);
 			// decompress the memory buffer (verbose=0)
 			ret = BZ2_bzBuffToBuffDecompress(DecompressedData, &outSize, FileData, inSize, 0, Verbosity);
 		} // while
@@ -1009,6 +1012,7 @@
 			return (NULL);
 		}
 
+		tern_lineup(0);
 		// compress the memory buffer (blocksize=9*100k, verbose=0, worklevel=30)
 		ret = BZ2_bzBuffToBuffCompress(CompressedData, &outSize, FileData, inSize, BWTblockSize, Verbosity, 30);
 		if (ret != BZ_OK)
@@ -1666,6 +1670,8 @@
 	}
 	pthread_mutex_init(MemMutex, NULL);
 
+	tern_lineup_init(0, numCPU, 20*numCPU);
+
 	// create queue
 	fifo = queueInit(numCPU);
 	if (fifo == NULL)
