diff -ruN pfscan-1.0/Makefile pfscan-1.0-modified/Makefile
--- pfscan-1.0/Makefile	2002-03-25 17:51:21.000000000 -0500
+++ pfscan-1.0-modified/Makefile	2013-02-18 21:48:36.561929316 -0500
@@ -23,9 +23,9 @@
 SOL_LIBS= -lpthread -lnsl -lsocket
 
 ## Linux 2.4 with Gcc 2.96
-LNX_CC=gcc -Wall -g -O
+LNX_CC=gcc -Wall -g -O -I$(XTERN_ROOT)/include
 LNX_LDOPTS=-Wl,-s 
-LNX_LIBS=-lpthread -lnsl
+LNX_LIBS=-lpthread -lnsl  -L$(XTERN_ROOT)/dync_hook -Wl,--rpath,$(XTERN_ROOT)/dync_hook -lxtern-annot
 
 
 OBJS = pfscan.o bm.o version.o pqueue.o
diff -ruN pfscan-1.0/pfscan.c pfscan-1.0-modified/pfscan.c
--- pfscan-1.0/pfscan.c	2002-03-25 18:13:45.000000000 -0500
+++ pfscan-1.0-modified/pfscan.c	2013-02-18 21:49:21.325927712 -0500
@@ -34,7 +34,7 @@
 
 #include "pqueue.h"
 #include "bm.h"
-
+#include "tern/user.h"
 
 
 extern char version[];
@@ -472,7 +472,9 @@
     
     while (pqueue_get(&pqb, (void **) &path) == 1)
     {
+    tern_non_det_start();
 	scan_file(path);
+    tern_non_det_end();
 	free(path);
     }
 
@@ -535,9 +537,11 @@
 
     nworkers = 2;
 
-    pthread_mutex_init(&print_lock, NULL);
     pthread_mutex_init(&aworker_lock, NULL);
+    tern_non_det_start();
+    pthread_mutex_init(&print_lock, NULL);
     pthread_mutex_init(&matches_lock, NULL);
+    tern_non_det_end();
 
     for (i = 1; i < argc && argv[i][0] == '-'; i++)
 	for (j = 1; j > 0 && argv[i][j]; ++j)
