--- NPB3.3.1-orig/NPB3.3-OMP/UA/transfer.f	2006-03-14 19:32:06.000000000 -0500
+++ NPB3.3.1/NPB3.3-OMP/UA/transfer.f	2013-02-26 23:32:13.668975503 -0500
@@ -5,6 +5,7 @@
 c------------------------------------------------------------------
 
       include 'header.h'
+      include 'tern/userF.h'
 
       integer i
 
@@ -261,6 +262,7 @@
 c------------------------------------------------------------------
 
       include 'header.h'
+      include 'tern/userF.h'
 
       double precision third
       parameter (third=1.d0/3.d0)
@@ -268,11 +270,11 @@
 
       double precision tmp,tmp1,tx(*),tmor(*),temp(lx1,lx1,2),
      &                 top(lx1,2)
-      integer il1,il2,il3,il4,ig1,ig2,ig3,ig4,ie,iface,nnje,
+      integer il1,il2,il3,il4,ig1,ig2,ig3,ig4,ie,iface,nnje,iii1,
      &        ije1,ije2,col,i,j,ije,ig,il
 
 c$OMP PARALLEL DEFAULT(SHARED) PRIVATE(il,j,ig,i,col,ije2,ije1,ig4,
-c$OMP& ig3,ig2,ig1,nnje,il4,il3,il2,il1,iface,ie,ije,
+c$OMP& ig3,ig2,ig1,nnje,il4,il3,il2,il1,iface,ie,ije,iii1,
 c$OMP& tmp,shift,temp,top,tmp1)
 
 c$OMP DO
@@ -282,6 +284,13 @@
 c$OMP END DO
 
 c$OMP DO
+      do iii1=1,2
+        call tern_non_det_start()
+      end do
+c$OMP END DO
+
+
+c$OMP DO
       do ie=1,nelt
         do iface=1,nsides
 c.........nnje=1 for conforming faces, nnje=2 for nonconforming faces
@@ -561,6 +570,13 @@
         end do
       end do
 c$OMP END DO NOWAIT
+
+c$OMP DO
+      do iii1=1,2
+        call tern_non_det_end()
+      end do
+c$OMP END DO
+
 c$OMP END PARALLEL 
 
       return
@@ -800,6 +816,7 @@
 c------------------------------------------------------------------------
 
       include 'header.h'
+      include 'tern/userF.h'
 
       double precision tx(lx1,lx1),tmor(lx1,lx1),bottom(lx1),
      &                 temp(lx1,lx1)
@@ -863,14 +880,15 @@
 c-------------------------------------------------------------------
 
       include 'header.h'
+      include 'tern/userF.h'
 
       double precision third
       parameter (third = 1.d0/3.d0)
       double precision tx(*)
-      integer il1,il2,il3,il4,ig1,ig2,ig3,ig4,ie,iface,col,j,ig,il
+      integer il1,il2,il3,il4,ig1,ig2,ig3,ig4,ie,iface,col,j,ig,il,iii2
 
 c$OMP PARALLEL DEFAULT(SHARED) PRIVATE(IE,IFACE,IL1,IL2,
-c$OMP& IL3,IL4,IG1,IG2,IG3,IG4,COL,J,IG,IL) 
+c$OMP& IL3,IL4,IG1,IG2,IG3,IG4,COL,J,IG,IL,iii2) 
 
 c$OMP DO
       do j=1,nmor
@@ -879,6 +897,12 @@
 c$OMP END DO
 
 c$OMP DO
+      do iii2=1,2
+        call tern_non_det_start()
+      end do
+c$OMP END DO
+
+c$OMP DO
       do ie=1,nelt
         do iface=1,nsides
           if(cbc(iface,ie).ne.3)then
@@ -965,6 +989,13 @@
         end do
       end do
 c$OMP END DO NOWAIT
+
+c$OMP DO
+      do iii2=1,2
+        call tern_non_det_end()
+      end do
+c$OMP END DO
+
 c$OMP END PARALLEL
       return
       end
@@ -978,14 +1009,15 @@
 c-------------------------------------------------------------------
 
       include 'header.h'
+      include 'tern/userF.h'
 
       double precision third
       parameter (third = 1.d0/3.d0)
       double precision tx(*)
-      integer il1,il2,il3,il4,ig1,ig2,ig3,ig4,ie,iface,col,j,ig,il
+      integer il1,il2,il3,il4,ig1,ig2,ig3,ig4,ie,iface,col,j,ig,il,iii3
 
 c$OMP PARALLEL DEFAULT(SHARED) PRIVATE(IE,IFACE,IL1,IL2,
-c$OMP& IL3,IL4,IG1,IG2,IG3,IG4,COL,J,IG,IL)
+c$OMP& IL3,IL4,IG1,IG2,IG3,IG4,COL,J,IG,IL,iii3)
 
 c$OMP DO     
       do j=1,nmor
@@ -998,10 +1030,15 @@
       end do
 c$OMP END DO
 
+c$OMP DO
+      do iii3=1,2
+        call tern_non_det_start()
+      end do
+c$OMP END DO
+
 c$OMP DO 
       do ie=1,nelt
         do iface=1,nsides
-          
           if(cbc(iface,ie).ne.3)then
             il1 = idel(1,  1,  iface,ie)
             il2 = idel(lx1,1,  iface,ie)
@@ -1095,6 +1132,13 @@
         end do
       end do
 c$OMP END DO NOWAIT
+
+c$OMP DO
+      do iii3=1,2
+        call tern_non_det_end()
+      end do
+c$OMP END DO
+
 c$OMP END PARALLEL
 
       return
