diff -Naur ImageMagick-6.8.0-4/magick/draw.c ImageMagick-6.8.0-4_modified/magick/draw.c
--- ImageMagick-6.8.0-4/magick/draw.c	2012-10-15 14:26:39.000000000 -0400
+++ ImageMagick-6.8.0-4_modified/magick/draw.c	2013-02-04 16:56:04.811873448 -0500
@@ -85,6 +85,7 @@
 #include "magick/token.h"
 #include "magick/transform.h"
 #include "magick/utility.h"
+#include "tern/user.h"
 
 /*
   Define declarations.
@@ -3957,6 +3958,7 @@
             status=MagickFalse;
             continue;
           }
+        tern_lineup(0);
         for ( ; x <= stop; x++)
         {
           if ((x == (ssize_t) ceil(primitive_info->point.x-0.5)) &&
@@ -4020,6 +4022,7 @@
         status=MagickFalse;
         continue;
       }
+    tern_lineup(0);
     for (x=start; x <= stop; x++)
     {
       /*
