--- racey/racey-orig.c	2013-03-09 22:28:55.483105782 -0500
+++ racey/racey.c	2013-03-09 22:53:24.683053214 -0500
@@ -126,14 +126,24 @@
    */
   
   for(i = 0 ; i < MAX_LOOP; i++) {
+    pthread_mutex_lock(&mutex);
     unsigned num = sig[threadId];
+    pthread_mutex_unlock(&mutex);
     unsigned index1 = num%MAX_ELEM;
     unsigned index2;
+    pthread_mutex_lock(&mutex);
     num = mix(num, m[index1].value);
+    pthread_mutex_unlock(&mutex);
     index2 = num%MAX_ELEM;
+    pthread_mutex_lock(&mutex);
     num = mix(num, m[index2].value);
+    pthread_mutex_unlock(&mutex);
+    pthread_mutex_lock(&mutex);
     m[index2].value = num;
+    pthread_mutex_unlock(&mutex);
+    pthread_mutex_lock(&mutex);
     sig[threadId] = num;
+    pthread_mutex_unlock(&mutex);
     /* Optionally, yield to other processors (Solaris use sched_yield()) */
     /* pthread_yield(); */
   }
